include/smear.h

```diff
@@ -24,6 +27,40 @@
         SRT_send_message(msg, CAT(m, _handler));                       \
     }
 
+/* cancel_token_t SRT_delayed_send(<machine name>, <event name>, void *msg,
+                                   uint64_t delay_ms); */
+/* Send an event with a delay in milliseconds. Returns a
+ * cancel_token_t that should be given back with a call to SRT_cancel
+ * later. This is a limited resource, so failure to release it will
+ * result in a resource leak.
+ * 
+ * When calling this, machine and event_name are the mangled parts of
+ * the event function you want invoked. For example, if you've got a
+ * state machine called `TURNSTILE` with an event called `timeout` and
+ * you want to send TURNSTILE_timeout(NULL) after 3 seconds, you would
+ * write
+ *
+ * timer_token = SRT_delayed_send(TURNSTILE, timeout, NULL, 3000);
+ *
+ * then, later, call
+ *
+ * SRT_cancel(timer_token);
+ */
+#define SRT_delayed_send(machine, event_name, msg, delay_ms)            \
+    ({                                                                  \
+        CAT(machine, _Event_Wrapper) *wrapper;                          \
+        wrapper = malloc(sizeof(CAT(machine, _Event_Wrapper)));         \
+        if (wrapper == NULL) SMUDGE_panic();                            \
+        wrapper->id = CAT4(EVID_, machine, _, event_name);              \
+        CAT(wrapper->event.e_, event_name) = msg;                       \
+        SRT_send_later(wrapper,                                         \
+                       (void (*)(const void *))                         \
+                       CAT3(machine, _, event_name),                    \
+                       delay_ms);                                       \
+    })
+
+typedef size_t cancel_token_t;
```

Comment:
I don't want to add a layer of indirection, and as long as people use the typedef it won't break stuff if we change it.

It's a size_t because it's an index into an array. But that's an implementation detail that nobody should care about.

Reply:
"As long as people" = famous last words.  I'd rather not have to rely on that.  I've seen this pattern (IDs represented as integer indices into an array) once before, and someone almost immediately began relying on the underlying representation.

Reply:
What about file handles? They're `int`s and I don't see a lot of arguments that they should have been pointers to opaque structs. 

I don't see how anyone could get anything out of it other than the size of the number. Users don't get access to the array (it's hidden behind an opaque struct pointer) and it's not like there's any useful arithmetic you can do on them.

Reply:
Fair point regarding file handles.  In fact, I would argue that file handles being integers is beneficial, because everyone knows the meaning of handles 1, 2, and 3.

In the case I'm thinking about, someone did two things: 1) dropped the typedeffed type, in favor of directly referring to the underlying type, which made it a bear to trace down when we wanted to fix it, and 2) started using them as indices into their own cache, which is a conceptual nightmare.

Reply:
Sounds like someone doesn't know how to write software.

Reply:
I can't disagree, and I am constantly paranoid of becoming beholden to such a person.  If we can prevent a dependency on the semantics of these tokens then I lean in that direction, just to avoid the rise of supporting it forever.

Reply:
My issue is I don't want to change my implementation in significant ways to accommodate horrible programmers.

The obvious way to do this is to use an array of structs and hand the users an opaque pointer to the struct holding their token. However, if I ever have to resize the array I'll be SOL because now I want to relocate these objects and I've given users pointers to them. That's not a fun rabbit hole.

I guess an alternative is to pretend it's a pointer and use intptr_t as my array index, and cast it inside my code. Do you think that would be preferable?

Honestly, though, if someone goes abusing the API like that I won't feel at all bad about breaking their code. I might feel bad for their co-workers. It's not like it's any harder to take an opaque pointer and cast it to `void *` and use it for nefarious purposes.